sql/core/src/test/scala/org/apache/spark/sql/SubquerySuite.scala

```diff
@@ -2598,4 +2598,29 @@ class SubquerySuite extends QueryTest
         Row("aa"))
     }
   }
+
+  test("SPARK-42346: Rewrite distinct aggregates after merging subqueries") {
+    withTempView("t1") {
+      Seq((1, 2), (3, 4)).toDF("c1", "c2").createOrReplaceTempView("t1")
+
+      checkAnswer(sql(
+        """
+          |SELECT
+          |  (SELECT count(distinct c1) FROM t1),
+          |  (SELECT count(distinct c2) FROM t1)
+          |""".stripMargin),
+        Row(2, 2))
```

Comment:
It seems `MergeScalarSubqueries` can not improve this case:
Spark 3.2:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [Subquery subquery#185, [id=#223] AS scalarsubquery()#198L, Subquery subquery#186, [id=#242] AS scalarsubquery()#199L]
   :  :- Subquery subquery#185, [id=#223]
   :  :  +- AdaptiveSparkPlan isFinalPlan=false
   :  :     +- HashAggregate(keys=[], functions=[count(distinct c1#153)], output=[count(DISTINCT c1)#193L])
   :  :        +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=221]
   :  :           +- HashAggregate(keys=[], functions=[partial_count(distinct c1#153)], output=[count#202L])
   :  :              +- HashAggregate(keys=[c1#153], functions=[], output=[c1#153])
   :  :                 +- Exchange hashpartitioning(c1#153, 200), ENSURE_REQUIREMENTS, [plan_id=217]
   :  :                    +- HashAggregate(keys=[c1#153], functions=[], output=[c1#153])
   :  :                       +- FileScan parquet default.t1[c1#153] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/Downloads/spark-3.2.3-bin-hadoop2.7/spark-warehous..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c1:int>
   :  +- Subquery subquery#186, [id=#242]
   :     +- AdaptiveSparkPlan isFinalPlan=false
   :        +- HashAggregate(keys=[], functions=[count(distinct c2#197)], output=[count(DISTINCT c2)#195L])
   :           +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=240]
   :              +- HashAggregate(keys=[], functions=[partial_count(distinct c2#197)], output=[count#206L])
   :                 +- HashAggregate(keys=[c2#197], functions=[], output=[c2#197])
   :                    +- Exchange hashpartitioning(c2#197, 200), ENSURE_REQUIREMENTS, [plan_id=236]
   :                       +- HashAggregate(keys=[c2#197], functions=[], output=[c2#197])
   :                          +- FileScan parquet default.t1[c2#197] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/Downloads/spark-3.2.3-bin-hadoop2.7/spark-warehous..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c2:int>
   +- Scan OneRowRelation[]
```

After `MergeScalarSubqueries` and this PR:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=false
+- Project [Subquery subquery#0, [id=#36].count(DISTINCT c1) AS scalarsubquery()#11L, Subquery subquery#1, [id=#63].count(DISTINCT c2) AS scalarsubquery()#12L]
   :  :- Subquery subquery#0, [id=#36]
   :  :  +- AdaptiveSparkPlan isFinalPlan=false
   :  :     +- Project [named_struct(count(DISTINCT c1), count(DISTINCT c1)#5L, count(DISTINCT c2), count(DISTINCT c2)#8L) AS mergedValue#17]
   :  :        +- HashAggregate(keys=[], functions=[count(spark_catalog.default.t1.c1#19), count(c2#20)], output=[count(DISTINCT c1)#5L, count(DISTINCT c2)#8L])
   :  :           +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=33]
   :  :              +- HashAggregate(keys=[], functions=[partial_count(spark_catalog.default.t1.c1#19) FILTER (WHERE (gid#18 = 1)), partial_count(c2#20) FILTER (WHERE (gid#18 = 2))], output=[count#23L, count#24L])
   :  :                 +- HashAggregate(keys=[spark_catalog.default.t1.c1#19, c2#20, gid#18], functions=[], output=[spark_catalog.default.t1.c1#19, c2#20, gid#18])
   :  :                    +- Exchange hashpartitioning(spark_catalog.default.t1.c1#19, c2#20, gid#18, 5), ENSURE_REQUIREMENTS, [plan_id=29]
   :  :                       +- HashAggregate(keys=[spark_catalog.default.t1.c1#19, c2#20, gid#18], functions=[], output=[spark_catalog.default.t1.c1#19, c2#20, gid#18])
   :  :                          +- Expand [[c1#2, null, 1], [null, c2#3, 2]], [spark_catalog.default.t1.c1#19, c2#20, gid#18]
   :  :                             +- FileScan parquet spark_catalog.default.t1[c1#2,c2#3] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/opensource/spark/spark-warehouse/org.apache.spark...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c1:int,c2:int>
   :  +- Subquery subquery#1, [id=#63]
   :     +- AdaptiveSparkPlan isFinalPlan=false
   :        +- Project [named_struct(count(DISTINCT c1), count(DISTINCT c1)#5L, count(DISTINCT c2), count(DISTINCT c2)#8L) AS mergedValue#17]
   :           +- HashAggregate(keys=[], functions=[count(spark_catalog.default.t1.c1#19), count(c2#20)], output=[count(DISTINCT c1)#5L, count(DISTINCT c2)#8L])
   :              +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=60]
   :                 +- HashAggregate(keys=[], functions=[partial_count(spark_catalog.default.t1.c1#19) FILTER (WHERE (gid#18 = 1)), partial_count(c2#20) FILTER (WHERE (gid#18 = 2))], output=[count#23L, count#24L])
   :                    +- HashAggregate(keys=[spark_catalog.default.t1.c1#19, c2#20, gid#18], functions=[], output=[spark_catalog.default.t1.c1#19, c2#20, gid#18])
   :                       +- Exchange hashpartitioning(spark_catalog.default.t1.c1#19, c2#20, gid#18, 5), ENSURE_REQUIREMENTS, [plan_id=56]
   :                          +- HashAggregate(keys=[spark_catalog.default.t1.c1#19, c2#20, gid#18], functions=[], output=[spark_catalog.default.t1.c1#19, c2#20, gid#18])
   :                             +- Expand [[c1#2, null, 1], [null, c2#3, 2]], [spark_catalog.default.t1.c1#19, c2#20, gid#18]
   :                                +- FileScan parquet spark_catalog.default.t1[c1#2,c2#3] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/yumwang/opensource/spark/spark-warehouse/org.apache.spark...., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<c1:int,c2:int>
   +- Scan OneRowRelation[]
```

Reply:
No, the 2 plans differ as expected. After `MergeScalarSubqueries` the 2 subqueries became identical and both contain `count(DISTINCT c1)` and `count(DISTINCT c2)` in `mergedValue`.
The final physical plan with the reuse makes the improvement more visible:
```
== Physical Plan ==
AdaptiveSparkPlan isFinalPlan=true
+- == Final Plan ==
   *(1) Project [Subquery subquery#251, [id=#117].count(DISTINCT c1) AS scalarsubquery()#259L, ReusedSubquery Subquery subquery#251, [id=#117].count(DISTINCT c2) AS scalarsubquery()#260L]
   :  :- Subquery subquery#251, [id=#117]
   :  :  +- AdaptiveSparkPlan isFinalPlan=true
         +- == Final Plan ==
            *(3) Project [named_struct(count(DISTINCT c1), count(DISTINCT c1)#254L, count(DISTINCT c2), count(DISTINCT c2)#256L) AS mergedValue#265]
            +- *(3) HashAggregate(keys=[], functions=[count(t1.c1#267), count(c2#268)], output=[count(DISTINCT c1)#254L, count(DISTINCT c2)#256L])
               +- ShuffleQueryStage 1
                  +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=228]
                     +- *(2) HashAggregate(keys=[], functions=[partial_count(t1.c1#267) FILTER (WHERE (gid#266 = 1)), partial_count(c2#268) FILTER (WHERE (gid#266 = 2))], output=[count#271L, count#272L])
                        +- *(2) HashAggregate(keys=[t1.c1#267, c2#268, gid#266], functions=[], output=[t1.c1#267, c2#268, gid#266])
                           +- AQEShuffleRead coalesced
                              +- ShuffleQueryStage 0
                                 +- Exchange hashpartitioning(t1.c1#267, c2#268, gid#266, 5), ENSURE_REQUIREMENTS, [plan_id=193]
                                    +- *(1) HashAggregate(keys=[t1.c1#267, c2#268, gid#266], functions=[], output=[t1.c1#267, c2#268, gid#266])
                                       +- *(1) Expand [[c1#247, null, 1], [null, c2#248, 2]], [t1.c1#267, c2#268, gid#266]
                                          +- *(1) Project [_1#242 AS c1#247, _2#243 AS c2#248]
                                             +- *(1) LocalTableScan [_1#242, _2#243]
         +- == Initial Plan ==
            Project [named_struct(count(DISTINCT c1), count(DISTINCT c1)#254L, count(DISTINCT c2), count(DISTINCT c2)#256L) AS mergedValue#265]
            +- HashAggregate(keys=[], functions=[count(t1.c1#267), count(c2#268)], output=[count(DISTINCT c1)#254L, count(DISTINCT c2)#256L])
               +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=114]
                  +- HashAggregate(keys=[], functions=[partial_count(t1.c1#267) FILTER (WHERE (gid#266 = 1)), partial_count(c2#268) FILTER (WHERE (gid#266 = 2))], output=[count#271L, count#272L])
                     +- HashAggregate(keys=[t1.c1#267, c2#268, gid#266], functions=[], output=[t1.c1#267, c2#268, gid#266])
                        +- Exchange hashpartitioning(t1.c1#267, c2#268, gid#266, 5), ENSURE_REQUIREMENTS, [plan_id=110]
                           +- HashAggregate(keys=[t1.c1#267, c2#268, gid#266], functions=[], output=[t1.c1#267, c2#268, gid#266])
                              +- Expand [[c1#247, null, 1], [null, c2#248, 2]], [t1.c1#267, c2#268, gid#266]
                                 +- Project [_1#242 AS c1#247, _2#243 AS c2#248]
                                    +- LocalTableScan [_1#242, _2#243]
   :  +- ReusedSubquery Subquery subquery#251, [id=#117]
   +- *(1) Scan OneRowRelation[]
+- == Initial Plan ==
...
```